### DISCUSSION TITLE
Is running untrusted code that has been compiled from source with limited references safe?

### DISCUSSION THREAD

**Discussion by copygirl** [dotnet/roslyn#68289] [Q&A] [answered]
When creating a `CSharpCompilation`, passing in untrusted source code, but limiting the `references` to known safe APIs and disabling `unsafe`, how safe is it to run the resulting assembly? Are there any additional checks I could or should be making before running the code that I compiled myself?

### Context

I'm writing a game engine in C# in which one of the main goals was allowing runtime editing of code (written in C#), even in a multiplayer context, as well as pushing code to clients for them to execute. To make this possible, I was hoping compiling from source with restrictions would allow me to do this in a safe way for players.

I do not want to lose out on the performance benefits I get from direct memory access, especially since I'm using the C library [Flecs](https://github.com/SanderMertens/flecs) that I structured everything around. The archetype-based ECS design promises a lot of performance, that I would lose out on if I had to give up on this idea and give in to full-on sandboxing the untrusted code in a separate process or so.

I've already heard it "should" be [relatively safe to compile (but not run) untrusted code.](https://github.com/dotnet/roslyn/discussions/56347)

There is an [unanswered StackOverflow question](https://stackoverflow.com/questions/59594525/how-vulnerable-is-running-untrusted-code-with-limiting-references-using-assembly) from 3 years ago that was about the same topic.

Of course, actual security experts would have to look into this to confirm it with any amount of certainty. However, for now I'd be satisfied if some people knowledgeable with the topic could confirm this "should" be safe in theory. Otherwise, I don't see a lot of promise in continuing to pursue the implementation I had in mind for my project.

**Comment by CyrusNajmabadi** [comment]
You should consider it as safe as whatever access/capabilities that user/process has.  If it want security around this, then you should have an explicit external sandbox.

**Comment by jasonmalinowski** [comment]
The usual challenge here is defining precisely what you consider to be "safe" APIs -- things like "just don't allow file I/O" fail because of APIs like Activator.CreateInstance() where you can create anything on the fly.

You of course don't need an APIs at all to create infinite loops, allocate memory until your machine runs of memory, or a stack overflow.

**Reply by copygirl** [reply]
`System.Runtime` and anything else relating to reflection are obviously out of the question.

Not being able to prevent resource exhaustion and (for example) destroying the game's state is something I'm willing to pay. I just want to make sure that a player's computer is safe. I'm hoping I can protect the game from accidental errors bringing down the whole application by running the mod code in threads, but that's about it.

**Reply by CyrusNajmabadi** [reply]
Basically, in this approach, assume the code can do whatever the user/process can do. So that includes accessing all user data, exfiltrating it, and/or destroying it.

**Reply by copygirl** [reply]
@CyrusNajmabadi I do understand this approach to security, but I do want to understand why this is the case. Is C# written in such a way that there's no way to influence the compilation to not result in potentially dangerous code, even if I limit the functions it has access to? Why? What are the ways to get around this restriction?

**Reply by CyrusNajmabadi** [reply]
>  Is C# written in such a way that there's no way to influence the compilation to not result in potentially dangerous code

C# is written using normal windows/linux/mac OS/process semantics.  The semantics of this are that the process has privileges based on the user.  

> even if I limit the functions it has access to?

Because there's no particular enforcement of that, and it's trivial to break out of this.  We could invest in internal sandboxing... but why bother, when external sandboxes are already a thing and solve this problem with a much better track record.

> Why?

So that applications can do things for users on their behalf that the user wants.  That's just the way processes have been designed in operating systems for 50+ years.  There's been some movement here in the mobile space.  But ultimately, users want apps to do things they can do, so apps can do that.  

There has been a space for limiting that.  But either every app/runtime has to solve this.  Or you solve this once externally with a dedicated sandboxing system provided by the outer system/OS.  Now you have that system for all apps that can work uniformly.

> What are the ways to get around this restriction?

There is no actual protection for the restrictions you're talking about.  The runtime is replete with methods that could be used to do arbitrary memory corruption (and thus potentially cause anything to happen).  You can attempt to prevent all of these vectors, but you will likely never succeed, and so you must assume that any of those approaches will always fail and the user (and their data) is at risk.  If your threat model wants to address that, an external sandbox is needed.

**Reply by copygirl** [reply]
> C# is written using normal windows/linux/mac OS/process semantics. The semantics of this are that the process has privileges based on the user.

Hypothetically, let's imagine I write my own scripting language that compiles, say to C#, or to machine code. If I were to limit what you can write, I could likely make it safe. But I suppose C# is much too complicated for this approach to apply to it. I had hoped I could strip things enough to just "make it work".

> The runtime is replete with methods that could be used to do arbitrary memory corruption (and thus potentially cause anything to happen).

Isn't the point I'm making to restrict access to any such methods? And when you say "anything" - isn't that also limited? One can't just write to executable memory to run anything they want. At the very least you need to trick existing code to do something unintended. Or is that different with .NET?

I'm still curious about what a possible attack could look like, but I think you've got me convinced enough that I can see there's likely a lot of strange ways to get around the restrictions I had in mind.

----

But it sounds like I won't be able to achieve what I want how I had imagined. The next best option would probably be to sandbox as much of the game as possible and have a privileged process manage everything sensitive.

**Reply by CyrusNajmabadi** [reply, marked as answer]
> Hypothetically, let's imagine I write my own scripting language that compiles, say to C#, or to machine code. If I were to limit what you can write, I could likely make it safe.

Many have tried, and this often turns out to not be the case.  In any event, you are rebuilding a sandbox, and now have to do all the work to ensure that you truly *are* safe.  This includes things like hardening yourself against things like 'timing attacks' that can exfiltrate user data (like passwords).  This is a *very* hard domain, and just blocking functions is insufficient.  It's a much better idea to leave this to systems designed for this purpose, that are battle tested and continually updated as the threats evolve.

> But I suppose C# is much too complicated for this approach to apply to it.

It's not a matter of the language.  It's just that security is far more than "don't allow programs to call certain functions".

> Isn't the point I'm making to restrict access to any such methods?

1. You'd have to do that perfectly.  Across the entire surface area of the language, and the entire API surface area of .net.
2. This is only one class of security issue.  Data exfiltration through timing attacks is another.  See heartbleed, etc. 
3. You could certainly try to build this, but you're having to redo all the work and effort put in by teams of security specialists who have to make things safe from an ever expanding, and extremely diverse set of threats.  If you feel that's a worthwhile expenditure of your time, go for it.  It's not for us.  We would certainly do a worse job, and give (at best) an illusion of security, which would actually be worse than worthless.

> One can't just write to executable memory to run anything they want.

Why not?  What is stopping an app from doing that?  That's how JITs work after all.  They emit code, and then execute it.  

> At the very least you need to trick existing code to do something unintended.

Of course.  But that's the point here... you're trying to run *untrusted* code.  It sounded like you wanted to make that safe.  If you have untrusted code, you're going to have to assume it's going to be malicious and trying to do bad things.  Tricking the system into doing unintended things would be one of the first things this code woudl try to do.

> The next best option would probably be to sandbox as much of the game as possible and have a privileged process manage everything sensitive.

Yes.  Just do this.
